### PR TITLE
[Spec] Ensures for staking_contract

### PR DESCRIPTION
**Pull request by UIZorrot**
### Description

This is the high-level properties and the ensures for staking_contract module.

### Test Plan

aptos move prove -f staking_contract

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move [hunk 1]
```diff
@@ -245,13 +306,12 @@ spec aptos_framework::staking_contract {
...
-        // TODO: this function is normal in the local machine
-        // However, timeout in the github test
-
-        pragma verify = false;
-
+        // preconditions
         include stake::ResourceRequirement;
+
         let staker_address = signer::address_of(staker);
+        // verify account::create_resource_account()
```
Comment (junkil-park) [inline aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:313]: Could you rephrase the comment "// verify XYZ" with "// postconditions of XYZ"? (e.g., `// postconditions of account::create_resource_account()`) And, do the same for the other places?

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move [hunk 2]
```diff
@@ -237,6 +287,17 @@ spec aptos_framework::staking_contract {
         // TODO: set because of timeout
         pragma verify = false;
         include GetStakingContractAmountsAbortsIf;
+
+        let pool_address = staking_contract.pool_address;
```
Comment (junkil-park) [inline aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:291]: The "verify" pragma is set to false. Are these specs tested?

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move [hunk 3]
```diff
@@ -237,6 +287,17 @@ spec aptos_framework::staking_contract {
         // TODO: set because of timeout
         pragma verify = false;
```
Comment (junkil-park) [inline aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:288]: Set "aborts_if_is_partial"?

**Review by rahxephon89 — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move [hunk 4]
```diff
@@ -245,13 +306,12 @@ spec aptos_framework::staking_contract {
         voter: address,
         contract_creation_seed: vector<u8>,
     ): (signer, SignerCapability, OwnerCapability) {
-        // TODO: this function is normal in the local machine
-        // However, timeout in the github test
-
-        pragma verify = false;
-
+        // preconditions
```
Comment (rahxephon89) [inline aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:309]: Please remove this comment

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move [hunk 5]
```diff
@@ -352,17 +421,127 @@ spec aptos_framework::staking_contract {
...
+        // verify stake::add_stake_with_cap()
+        include StakeAddStakeWithCapAbortsIf{ pool_address: resource_addr };
+
+        let post post_store = global<Store>(staker_address);
+        let post post_staking_contracts = post_store.staking_contracts;
+        // property 2: A staking contract is created and stored in a mapping within the Store resource.
+        ensures simple_map::spec_contains_key(post_staking_contracts, operator);
+    }
+
+    spec schema StakeAddStakeWithCapAbortsIf {
```
Comment (rahxephon89) [inline aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:487]: As mentioned before, if a schema contains both aborts_if and ensures, make sure use a proper name.

Comment (rahxephon89) [reply aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:487]: I think you have similar spec in the stake spec PR. Is it possible make it more modular by landing the stake module PR first and use the schema there?

**Review by rahxephon89 — CHANGES_REQUESTED** [review]

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move (hunk 5 above)
Comment (UIZorrot) [reply aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:487]: I'll try that later

**Review by rahxephon89 — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move (hunk 1 above)
Comment (UIZorrot) [reply aptos-move/framework/aptos-framework/sources/staking_contract.spec.move]: settled

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move (hunk 2 above)
Comment (UIZorrot) [reply aptos-move/framework/aptos-framework/sources/staking_contract.spec.move:283]: yes, it has been tested and it passed, remove thhe verify

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move [hunk 6]
```diff
@@ -1,27 +1,36 @@
 spec aptos_framework::staking_contract {
+    /// We cannot complete high level properties 6 to 8, because we cannot verify new cycles.
```
[the remark this answers is not in the document]
Comment (UIZorrot) [reply aptos-move/framework/aptos-framework/sources/staking_contract.spec.move]: removed

File: aptos-move/framework/aptos-framework/sources/staking_contract.spec.move (hunk 4 above)
Comment (UIZorrot) [reply aptos-move/framework/aptos-framework/sources/staking_contract.spec.move]: removed

**Review by junkil-park — APPROVED** [review]
